### PR TITLE
[RFC] Proposal to Update PPO Test to Add LR Scheduler

### PR DESCRIPTION
**Pull request by Seoley**
#### Context
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

#### Changelog
Related Issue: https://github.com/pytorch/torchtune/issues/2064#issuecomment-2668489405

I refered to https://github.com/pytorch/torchtune/issues/2263 and add an lr_scheduler option to the Mistral PPO recipe. Then, run the recipe using the following command:
`tune run ppo_full_finetune_single_device --config mistral/7B_full_ppo_low_memory`

Here are the loss curves for reference:
PPO_loss_curve

And then, I need to update the test code in `recipes/test_ppo_full_finetune_single_device.py`

This commit includes changes where I added the lr_scheduler option to recipes/test_ppo_full_finetune_single_device.py, but the test has failed. I suspect the failure might be due to using LLaMA.
How should I resolve this issue? I would like to ask for some advice.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [X] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Review by joecummings — COMMENTED** [review]
@SalmanMohammadi  Can you take a look?

File: recipes/ppo_full_finetune_single_device.py [hunk 1]
```diff
@@ -257,6 +258,18 @@ def setup(self, cfg: DictConfig) -> None:
             * (self.batch_size // self._ppo_batch_size)
         )
 
+        self._steps_per_epoch = (
+            len(self._dataloader) // self._gradient_accumulation_steps
+        )
+        self.global_step = self._epochs_run * self._steps_per_epoch
```
Comment (SalmanMohammadi) [inline recipes/ppo_full_finetune_single_device.py:264]: `global_step` is already defined here https://github.com/pytorch/torchtune/blob/80da6a5dae23a201595d07041c12ffde830332d7/recipes/ppo_full_finetune_single_device.py#L253 - how come you're re-defining it here?

File: recipes/ppo_full_finetune_single_device.py [hunk 2]
```diff
@@ -257,6 +258,18 @@ def setup(self, cfg: DictConfig) -> None:
             * (self.batch_size // self._ppo_batch_size)
         )
 
+        self._steps_per_epoch = (
```
Comment (SalmanMohammadi) [inline recipes/ppo_full_finetune_single_device.py:261]: I think we should name this `lr_steps` and the correct value would be the total number of optimizer steps being taken, which should be `self._total_steps * self._ppo_epochs * (self.batch_size // self._ppo_batch_size)`, right?

**Comment by SalmanMohammadi** [comment]
Hi @Seoley! Thanks so much for taking this on. Sorry for the delay in getting round to a review here.

I think this generally looks good - I think the only comments I have are how we calculate the number of steps for the scheduler, which I've left inline. Let me know what you think : )

**Comment by Seoley** [comment]
Hi @SalmanMohammadi! Thanks for the review. Sorry for the late reply, I was on vacation. I will check and reply this weekend!

**Comment by SalmanMohammadi** [comment]
Hi @Seoley - let me know if you still plan on finishing this, otherwise I'm happy to help get it over the line.

**Comment by Seoley** [comment]
Hi @SalmanMohammadi
Sorry for the delay, I’ve been a bit tied up recently so I lost track of this thread. I'm looking into it now and will get back to you soon! I haven’t had the chance to fully look into this yet, but it’s still on my radar. 
Thank you, as always, for your support!

File: recipes/ppo_full_finetune_single_device.py (hunk 1 above)
Comment (Seoley) [reply recipes/ppo_full_finetune_single_device.py:264]: You're right. I had redefined `global_step` in `full_finetune_single_device.py` to use it, but it seems I wrote it incorrectly. I've deleted the redefined code and I'm testing code now.

File: recipes/ppo_full_finetune_single_device.py (hunk 2 above)
Comment (Seoley) [reply recipes/ppo_full_finetune_single_device.py:261]: You're absolutely right! My code didn't properly reflect the structure of PPO. After your suggestion, I added `lr_step` and I'm also updating the definition of the `lr_scheduler` as follows:

```python
# one "step" is a single gradient update over a minibatch of trajectories
self.global_step = (
    self._steps_run
    * self._ppo_epochs
    * (self.batch_size // self._ppo_batch_size)
)

self.lr_step = self._total_steps * self._ppo_epochs * (self.batch_size // self._ppo_batch_size)

# Setup lr scheduler
self._lr_scheduler = self._setup_lr_scheduler(
    cfg_lr_scheduler=cfg.get("lr_scheduler", None),
    num_training_steps=self.lr_step,
    last_epoch=self.global_step - 1,
)
```
Did I understand your suggestion correctly?

Comment (SalmanMohammadi) [reply recipes/ppo_full_finetune_single_device.py:261]: This sounds good to me!

**Comment by Seoley** [comment]
@SalmanMohammadi I updated lr_step and global_step as you suggested and pushed the changes again.
Also, I accidentally wrote the wrong commit message while applying some formatting changes. Thanks for your understanding.

File: recipes/ppo_full_finetune_single_device.py [hunk 3]
```diff
@@ -257,6 +258,19 @@ def setup(self, cfg: DictConfig) -> None:
...
+        self.lr_step = (
+            self._total_steps
+            * self._ppo_epochs
+            * (self.batch_size // self._ppo_batch_size)
+        )
+
+        # Setup lr scheduler
+        self._lr_scheduler = self._setup_lr_scheduler(
+            cfg_lr_scheduler=cfg.get("lr_scheduler", None),
+            num_training_steps=self.lr_step,
```
Comment (SalmanMohammadi) [inline recipes/ppo_full_finetune_single_device.py:270]:
```suggestion
            num_training_steps=lr_step,
```

File: recipes/ppo_full_finetune_single_device.py [hunk 4]
```diff
@@ -257,6 +258,19 @@ def setup(self, cfg: DictConfig) -> None:
             * (self.batch_size // self._ppo_batch_size)
         )
 
+       lr_steps = (
```
Comment (SalmanMohammadi) [inline recipes/ppo_full_finetune_single_device.py:261]:
```suggestion
        lr_steps = (
```

**Review by SalmanMohammadi — APPROVED** [review]
